File: backend/app/parsers/manual.py
"""Manual asset entry — PPF, FD, SGB, NPS, bonds, real estate, cash, digital gold, custom.

These rarely have ISINs or live prices, so they're value-based: the user states the current
value (and optionally invested value). A sensible InstrumentType is required so the classifier
can bucket them (e.g. PPF -> Debt/PPF, SGB -> Gold/SGB).
"""

from __future__ import annotations

from datetime import date as date_cls

from pydantic import BaseModel

from app.domain.taxonomy import InstrumentType, Source
from app.parsers.base import ParsedHolding, ParsedTxn, ParseResult


class ManualEntry(BaseModel):
    name: str
    instrument_type: InstrumentType
    current_value: float
    invested_value: float | None = None
    quantity: float = 1.0
    isin: str | None = None
    symbol: str | None = None
    account_name: str = "Manual"
    institution: str | None = None
    sector: str | None = None
    market_cap: str | None = None
    asset_class: str | None = None  # explicit override hint
    # Optional purchase/deposit date → one dated cashflow so the asset gets an XIRR.
    # Accurate for lump-sum assets (FD, SGB, bonds); approximate for staggered ones (PPF/EPF/NPS).
    start_date: date_cls | None = None
    note: str | None = None


def parse_entries(entries: list[ManualEntry], file_name: str | None = None) -> ParseResult:
    result = ParseResult(source=Source.MANUAL, file_name=file_name)
    for e in entries:
        kind = "bank" if e.instrument_type in (InstrumentType.FD, InstrumentType.CASH) else "manual"
        invested = e.invested_value if e.invested_value is not None else e.current_value
        txns: list[ParsedTxn] = []
        if e.start_date and invested:
            txns.append(ParsedTxn(date=e.start_date, kind="buy", amount=round(-invested, 2)))
        result.holdings.append(
            ParsedHolding(
                name=e.name,
                instrument_type=e.instrument_type,
                isin=(e.isin or "").upper() or None,
                symbol=(e.symbol or "").upper() or None,
                quantity=e.quantity,
                current_value=e.current_value,
                invested_value=invested,
                source=Source.MANUAL,
                account_name=e.account_name,
                account_kind=kind,
                institution=e.institution,
                sector_hint=e.sector,
                market_cap_hint=e.market_cap,
                category_hint=e.asset_class,
                transactions=txns,
                raw={"note": e.note} if e.note else {},
            )
        )
    result.info(f"Added {len(result.holdings)} manual entries.")
    return result

File: frontend/src/engine/parse/manual.ts
// Manual asset entry → TS mirror of backend/app/parsers/manual.py.

import { InstrumentType, Source } from "../taxonomy";
import { ParsedTxn, ParseResult, holding } from "./types";
import { round2 } from "./csv";

export interface ManualEntry {
  name: string;
  instrument_type: string;
  current_value: number;
  invested_value?: number | null;
  quantity?: number;
  isin?: string | null;
  symbol?: string | null;
  account_name?: string;
  institution?: string | null;
  sector?: string | null;
  market_cap?: string | null;
  asset_class?: string | null;
  start_date?: string | null; // ISO YYYY-MM-DD
  note?: string | null;
}

export function parseManualEntries(entries: ManualEntry[], fileName: string | null = "manual-entry"): ParseResult {
  const result = new ParseResult(Source.MANUAL, fileName);
  for (const e of entries) {
    const kind = e.instrument_type === InstrumentType.FD || e.instrument_type === InstrumentType.CASH ? "bank" : "manual";
    const invested = e.invested_value != null ? e.invested_value : e.current_value;
    const txns: ParsedTxn[] = [];
    if (e.start_date && invested) txns.push({ date: e.start_date, kind: "buy", amount: round2(-invested) });
    result.holdings.push(holding({
      name: e.name,
      instrument_type: e.instrument_type,
      isin: (e.isin || "").toUpperCase() || null,
      symbol: (e.symbol || "").toUpperCase() || null,
      quantity: e.quantity ?? 1.0,
      current_value: e.current_value,
      invested_value: invested,
      source: Source.MANUAL,
      account_name: e.account_name ?? "Manual",
      account_kind: kind,
      institution: e.institution ?? null,
      sector_hint: e.sector ?? null,
      market_cap_hint: e.market_cap ?? null,
      category_hint: e.asset_class ?? null,
      transactions: txns,
      raw: e.note ? { note: e.note } : {},
    }));
  }
  result.info(`Added ${result.holdings.length} manual entries.`);
  return result;
}
